Add ModifyElementSubdomainByVar UserObject

dewenyushu opened the pull request:
Address #27486

dewenyushu committed:
Add ModifyElementSubdomainByVar, refs #27486

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2196947/) on 73cc2a9 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27487/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2197483/) on 7b99d26 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>53c560</th>
      <th colspan=4>#27487 7b99d2</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.13%</td>
      <td>85.13%</td>
      <td>+0.00%</td>
      <td>95.24%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>103068</td>
      <td>103087</td>
      <td>+19</td>
      <td>20</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18004</td>
      <td>18006</td>
      <td>+2</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27487/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27487/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27487/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on framework/include/userobjects/ModifyElementSubdomainByVar.h:
```diff
@@ -0,0 +1,27 @@
+#include "ElementSubdomainModifier.h"
+#include "MooseEnum.h"
+
+class ModifyElementSubdomainByVar : public ElementSubdomainModifier
```
docstring

GiudGiud commented on framework/src/userobjects/ModifyElementSubdomainByVar.C:
```diff
@@ -0,0 +1,59 @@
+{
+  // Calculate the desired subdomain ID for the current element.
+  Real val = 0.0;
+  for (unsigned int qp = 0; qp < _qrule->n_points(); ++qp)
```
```suggestion
  for (const auto qp : make_range(_qrule->n_points()))
```

GiudGiud commented on framework/src/userobjects/ModifyElementSubdomainByVar.C:
```diff
@@ -0,0 +1,59 @@
+    auto it = _mesh.meshSubdomains().lower_bound(sid);
+    SubdomainID nearest_id =
+        it == _mesh.meshSubdomains().end() ? *_mesh.meshSubdomains().rbegin() : *it;
+    mooseDoOnce(mooseWarning(
```
We dont like to do this anymore.

I suggest you keep a vector (with a lock for threads?) that keeps track of which subdomains you have output a warning for, to output the warning only once per subdomain

dewenyushu committed:
Address comments

dewenyushu committed:
Add lock

dewenyushu commented on framework/src/userobjects/ModifyElementSubdomainByVar.C:
```diff
@@ -0,0 +1,66 @@
+    SubdomainID nearest_id =
+        it == _mesh.meshSubdomains().end() ? *_mesh.meshSubdomains().rbegin() : *it;
+
+    // Store the target subdomain ID if it hasn't been previously requested.
+    // Lock the _void_sids for thread-safe operations.
+    std::lock_guard<std::mutex> lock(_void_sids_mutex);
+    if (_void_sids.find(sid) == _void_sids.end())
+    {
+      mooseWarning("Requested subdomain ",
+                   sid,
+                   " does not exist. Subdomain ID ",
+                   nearest_id,
+                   " is assigned. Please ensure the passed variable falls within the range of the "
+                   "mesh's subdomain IDs for the expected behavior.");
+      _void_sids.insert(sid);
+    }
+    return nearest_id;
```
@GiudGiud thanks for the comment about adding a lock for thread-safe operations. I am guessing this is the correct way of doing it?

GiudGiud replied:
looks good to me

dewenyushu commented:
@GiudGiud this should be ready :)

GiudGiud commented on framework/src/userobjects/ModifyElementSubdomainByVar.C:
```diff
@@ -0,0 +1,66 @@
+  // target subdomain ID does not exist in the mesh.
+  if (_mesh.meshSubdomains().find(sid) == _mesh.meshSubdomains().end())
+  {
+    auto it = _mesh.meshSubdomains().lower_bound(sid);
+    SubdomainID nearest_id =
+        it == _mesh.meshSubdomains().end() ? *_mesh.meshSubdomains().rbegin() : *it;
```
this isnt the nearest in a rounding sense? 
https://cplusplus.com/reference/set/set/lower_bound/
This is the one right below, with an adaptation if the end is found (so sid is bigger than everything?)

either change the wording in the doco to describe in what sense you mean closest, or compare to the next element to find the closest

dewenyushu replied:
Yeah good point. The lower bound does not equal to closest. It returns the pointer to the smallest value that is equal or greater than the inquired value. I just changed the naming and the doc to reflect this

dewenyushu committed:
Address more comments

dewenyushu commented:
@GiudGiud this should be ready (again)

hugary1995 commented:
I'd suggest a different name. To have slightly more consistency with the existing `CoupledVarThresholdElementSubdomainModifier`, this one could be named `VariableValueElementSubdomainModifier` which might also be more self-explanatory.

dewenyushu committed:
Change object name

Address more comments

Update test/tests/userobjects/element_subdomain_modifier/tests

dewenyushu commented:
> I'd suggest a different name. To have slightly more consistency with the existing `CoupledVarThresholdElementSubdomainModifier`, this one could be named `VariableValueElementSubdomainModifier` which might also be more self-explanatory.

Name changed!

GiudGiud commented on framework/src/userobjects/VariableValueElementSubdomainModifier.C:
```diff
@@ -0,0 +1,69 @@
+  SubdomainID sid = (unsigned int)(round(val));
+
+  // Verify whether the specified target subdomain ID is present within the mesh. If it is not
+  // found, locate the smallest subdomain ID in the mesh that matches or exceeds the target
+  // subdomain ID. Or select the largest subdomain ID present in the mesh if all subdomain IDs are
+  // smaller than the target.
```
basically this but in the doco

dewenyushu commented on framework/doc/content/source/userobjects/ModifyElementSubdomainByVar.md:
```diff
@@ -0,0 +1,12 @@
+
+The `ModifyElementSubdomainByVar` UserObject adjusts the element's subdomain according to a provided variable value. Subdomain IDs can be assigned based on the average variable value within the element, rounded to the nearest existing subdomain ID in the mesh.
+
+**Note:** If the desired value is not present in the subdomain ID list, a warning will be issued, and the closest existing subdomain ID value will be automatically assigned to the element.
```
Fixed!

GiudGiud commented on test/tests/userobjects/element_subdomain_modifier/tests:
```diff
@@ -1,6 +1,6 @@
-  design = 'CoupledVarThresholdElementSubdomainModifier.md'
+ [Tests]
+  issues = '#15795 #17100 #17239 #21383 #27486'
+  design = 'CoupledVarThresholdElementSubdomainModifier.md ModifyElementSubdomainByVar.md'
```
```suggestion
  design = 'CoupledVarThresholdElementSubdomainModifier.md VariableValueElementSubdomainModifier.md'
```
